### ISSUE TITLE
Keep connection open instead of polling for status

### ISSUE THREAD

**Issue by emontnemery** [TradeFace/tuya#29]
Would a PR keeping the connections to the devices instead of polling be accepted, or is there some reason for the polling mechanism?

**Comment by TradeFace** [comment]
The first version of the lib kept the connection(s) open. Some devices stopped sending replies after a few requests. Maybe this change wasn't necessary. I did a lot of changes to the code at the time to prevent the "json obj data unvalid" error. 

So sure I am open any PR that brings value to the lib. Please target it at the dev branch, so I can manually test if everything keeps working in absence of test-code ;)

**Comment by jkerdreux-imt** [comment]
Most devices I tested (something like 8 devices from various vendors) 
closes connections after 10 seconds.
All closes the connections on Smart Like request (ie turn on lights) and 
some doesn't even respond if this software
is open.

Most devices reset the connection on subsequent request..

Bye

**Comment by TradeFace** [comment]
So, tying to keep it open won't help... come to think of it. The connection isn't actively closed, that potentially could give issues when you try to poll it at high rate (or an other application)

**Comment by emontnemery** [comment]
`json obj data unvalid` seems to be a limitation in some devices, at least in my case it's triggered also by the official apps.

The reason I'd like to keep the connection open is that the devices will push a status update if it's turned on or off by local operation.

@jkerdreux-imt do you mean the connection is closed even with correct exchange of ping-pong messages?

@TradeFace I'll give it a try. Do you have any opinion on how to implement it, should each device have it's own worker thread reading from the connection + keeping the connection active or do you prefer some other way?

**Comment by TradeFace** [comment]
in [TuyaMQTT](https://github.com/TradeFace/tuyamqtt/) I use a thread per device indeed. That seems to work fine. There I have the polling frequency set at 5s. I did some preparations to be able to change it (polling frequency) on a per device basis, but didn't get around finishing it. 

I don't know how the original app gets the near instant updates of the device. I don't remember any data push from the device directly to the app... perhaps it sends you data to china first? :)

**Comment by emontnemery** [comment]
Both local and cloud connection is based on keeping connection to either the cloud or to the devices open, hence the near instant updates.

I'll go ahead and try a thread based implementation.

**Comment by jkerdreux-imt** [comment]
If the device doesn't receive any request in 10 seconds, it drops the 
connection.
I'm quite sure that asking the device state every 2 seconds will kill 
the connection too.
I own some Alfawise devices, and if you query them too much (says every 
20 sec), they will reboot once a hour.

But please don't introduce some threads in the API itself. Threads 
aren't the best way you handle concurrency in Python.
There is a lot of options (coroutine, gevent, greenlet, asyncio..). You 
should keep the plain functional API, and let
the end developer choose his best option.

Bye.

**Comment by emontnemery** [comment]
> There is a lot of options (coroutine, gevent, greenlet, asyncio..).

That's a good point, but the current implementation is based on blocking IO using `socket` so it does not really work with `asyncio` etc.

Here's an attempt at an *optional* helper thread to keep track of the connection: TradeFace/tuya#31

**Comment by jkerdreux-imt** [comment]
It works with gevent, simply spawn the call :) (it should be ok with 
coroutine, greenlet, stackless too)

Here my code :

https://redmine.telecom-bretagne.eu/projects/xaal/repository/entry/code/Python/branches/0.7/devices/protocols/Tuya/xaal/tuya/devices.py

Thanks

**Comment by jkerdreux-imt** [comment]
I tested the thread client this morning. It works as expected. Everything is fine until you run the Smart Life app. This cause the connection to reset, and the thread loop never manage to connect back. It loops on a connection reset. 

```
TuyaClient: reconnecting
TuyaClient: reconnecting
TuyaClient: Unexpected exception:
Traceback (most recent call last):
  File "/home/jkx/tuya/tuyaface/__init__.py", line 98, in run
    data = self.connection.recv(4096)
ConnectionResetError: [Errno 104] Connection reset by peer
TuyaClient: reconnecting
TuyaClient: reconnecting
TuyaClient: Unexpected exception:
Traceback (most recent call last):
  File "/home/jkx/tuya/tuyaface/__init__.py", line 98, in run
    data = self.connection.recv(4096)
ConnectionResetError: [Errno 104] Connection reset by peer
```

Hope this helps.

**Comment by TradeFace** [comment]
Yes it does help, much appreciated! So, sadly we can conclude this setup won't work in all instances. I am not familiar with gevent spawning as you suggested. Can that add value to the module?

**Comment by TradeFace** [comment]
I suggest we partially revert https://github.com/TradeFace/tuya/pull/31 by removing the TuyaClient class (and come up with another solution to up the status frequency, spawning perhaps) . The sequence number and passing connection are useful additions IMO. What do you guys think?

**Comment by emontnemery** [comment]
@TradeFace Sorry if I was unclear in #31.
By "WIP" I meant "Work In Progress", it was not quite ready for merge yet.

I think it's fully expected that the connection with the Tuya device is severed if the offical app is used, but the TuyaClient should reconnect of course.

The problem found by @jkerdreux-imt is due to a bug in the reconnect logic, I'll fix it tonight together with your review comments and some of my own TODOs.

The "gevent spawning" mentioned by @jkerdreux-imt is just another way of "hiding" a blocking function call, it doesn't really help in this case.

**Comment by TradeFace** [comment]
I Know, and I merged it anyway. Crazy! :) 

Alright, curious to see if you can get it solved

**Comment by emontnemery** [comment]
@jkerdreux-imt thanks for testing, I will fix the broken reconnect tonight!

Regarding this:
>Everything is fine until you run the Smart Life app. This cause the connection to reset,

Some Tuya devices only support 1 connection using the local API; do you know if that's the case for your devices? I guess a valid test would be to start a `TuyaClient` and then using the old non-threaded API to check if it also triggers the disconnect.

**Comment by jkerdreux-imt** [comment]
1) 
Most devices I tested produced this kind of error: Connection reset while the Smart Life App (SLApp)  is open. It does exactly the same w/ the old _send_request(). I'm just reporting the reconnect doesn't work, and should be fixed. Keep in mind that if you try to reconnect too quickly you we receive one more Connection reset. 

2) Gevent patch select(), time.sleep() .. to make them not-blocking anymore and so defer result. It allow you to write async code without the overhead of threading (which is a bad idea in Python due to GIL issues). I choose gevent but a lot of other stuff works like that, greenlet, stackless... 

3) No the old non-threaded API doesn't triggers a connection reset, but it is unable to connect due to ping-pong running. 2 ping-pong running in parallel will trigger the reconnect loop. (easy-to-try).

4) I think the select / connect / ping / pong stuff should be in the core API. And the ThreadClient should be provided too, but only consumming select / connect / ping-pong.. It should be easy to use the API without threads. I'm not in homeassistant team for example but they tend to avoid thread, and support mainly asyncio. perhaps an easy way to fix that, is to carry the connection in the device dict ? Not really programmer-friendly, but should be ok.

**Comment by TradeFace** [comment]
> perhaps an easy way to fix that, is to carry the connection in the device dict ? Not really programmer-friendly, but should be ok.

If you go that route then put the sequence number (and maybe others) in there as well.

**Comment by emontnemery** [comment]
The reconnect logic is fixed now, see PR #32 

It's a good idea and straight forward to keep the sequence number, and maybe the connection, in the device dict.

For a native HomeAssistant integration, a non blocking asyncio client using TradeFace/tuya and translating to HomeAssistant entities would needed. It does not matter if the underlying library, TradeFace/tuya in this case, is using threads or some other mechanism internally for parellelism.

Implementing HomeAssistant MQTT discovery in TradeFace/tuyamqtt would be easier than implementing a full integration.

**Comment by TradeFace** [comment]
> Implementing HomeAssistant MQTT discovery in TradeFace/tuyamqtt would be easier than implementing a full integration.

Generally agreed; direct integrations into HA bloats the system and makes you dependent of it. Better to "abuse" MQTT/Mosquitto as an event-bus. 

However there is an active direct (unstable) implementation [localtuya-homeassistant](https://github.com/mileperhour/localtuya-homeassistant/) Which uses the original pytuya module.

**Comment by TradeFace** [comment]
tuyamqtt isn't publishing the state.
status and set_state in tuyaclient aren't returning values.

**Comment by emontnemery** [comment]
Ok, I will take a look later today or tomorrow.

**Comment by TradeFace** [comment]
I'll fix the return issue in a minute. 

There is an uncaught exception in tuyaclient._connect. Maybe you can take a look at it later on.

**Comment by emontnemery** [comment]
Sure, can you paste a log with the exception?

**Comment by TradeFace** [comment]
WARNING:tuyaface:Failed to connect to 192.168.1.27. Retry in 1 seconds
ERROR:tuyaface.tuyaclient:TuyaClient: exception when opening socket
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tuyaface-1.1.7-py3.6.egg/tuyaface/tuyaclient.py", line 89, in run
    self._connect()
  File "/usr/local/lib/python3.6/dist-packages/tuyaface-1.1.7-py3.6.egg/tuyaface/tuyaclient.py", line 57, in _connect
    self.connection = _connect(self.device)
  File "/usr/local/lib/python3.6/dist-packages/tuyaface-1.1.7-py3.6.egg/tuyaface/__init__.py", line 246, in _connect
    raise e
  File "/usr/local/lib/python3.6/dist-packages/tuyaface-1.1.7-py3.6.egg/tuyaface/__init__.py", line 242, in _connect
    connection.connect((device['ip'], 6668))
socket.timeout: timed out

**Comment by emontnemery** [comment]
> it isn't uncaught, it just dumps a trace.

Ah, right. That was on purpose for debugging, but I guess there's no need to print the stacktrace for these kind of expected errors.
Should I open a PR to clean it up, or will you fix it?

> tuyamqtt isn't publishing the state.

Did you fix that already?

**Comment by TradeFace** [comment]
I can pass it. 

Yes it is now publishing status. 

I found an other problem for the special case devices (that return the "json obj data unvalid" error).
In my case a 3 button switch. When I switch all of them on simultaneously in HA only the last on is actually switched. I didn't look into it yet (best guess the command is overwritten)... but maybe it gives you an aha-moment

**Comment by emontnemery** [comment]
Do you mean not all MQTT commands sent by HA are handled?

**Comment by TradeFace** [comment]
It is hard to tell. It seems a bit random. Sometimes they all switch on and other times only dps 3 is switched. (this not with your latest PR)

_The neighbors must think I've officially gone mental, switching all lights in the house at the same time ;)_

**Comment by TradeFace** [comment]
```
ERROR:tuyaface.tuyaclient:TuyaClient: Unexpected exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tuyaface-1.1.7-py3.6.egg/tuyaface/tuyaclient.py", line 118, in run
    reply = json.loads(reply["data"])
  File "/usr/lib/python3.6/json/__init__.py", line 348, in loads
    'not {!r}'.format(s.__class__.__name__))
TypeError: the JSON object must be str, bytes or bytearray, not 'NoneType'
```

**Comment by TradeFace** [comment]
I just tested with your latest PR. It now reacts sluggish on HA commands.

**Comment by jkerdreux-imt** [comment]
Just test the development branch and Tuyaclient never call on_status now :(

**Comment by TradeFace** [comment]
I just reverted the PR for now.

**Comment by emontnemery** [comment]
@jkerdreux-imt Is the problem that even though TuyaClient is connected to a device and it receives status from the device, `on_status` is not called? Can you reproduce it and share a log?

**Comment by TradeFace** [comment]
I merged and did a quick test on https://github.com/TradeFace/tuya/pull/36 . I think (subjective) it got a little bit better, but it's still very slow to react; if it does at all. When I switch to complete set of switches some won't react at all. (apart from those out of reach)

And I have this exception
```
ERROR:tuyaface.tuyaclient:TuyaClient: Unexpected exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tuyaface-1.1.7-py3.6.egg/tuyaface/tuyaclient.py", line 122, in run
    reply = json.loads(reply["data"])
  File "/usr/lib/python3.6/json/__init__.py", line 348, in loads
    'not {!r}'.format(s.__class__.__name__))
TypeError: the JSON object must be str, bytes or bytearray, not 'NoneType'
```

**Comment by emontnemery** [comment]
OK, that's a shame!
I'll push an update with some additional logs to try to see what's going on at your side. Does this kind of rapid switching work well when `TuyaClient` is not used?

**Comment by emontnemery** [comment]
It would be awesome if you could test #37 also to make sure it doesn't break anything.

**Comment by TradeFace** [comment]
https://github.com/TradeFace/tuya/pull/37 seems to work much better. I'll compare it to the pre-tuyaclient setup in a minute. (after I check updates on manual device handling). I'll report back in a few min.

**Comment by TradeFace** [comment]
Indeed the rapid switching reacts as quick and at the same "error level" (due to local connection issues) as the pre-tuyaclient setup. Big gains in manual device handling status updates though! Well done!  :1st_place_medal: :100: 

The "Unexpected exception" remains though

**Comment by emontnemery** [comment]
> Indeed the rapid switching reacts as quick and at the same "error level" (due to local connection issues) as the pre-tuyaclient setup.

Great! Then #38 is not needed, just close it without merging.

> The "Unexpected exception" remains though

There's a quick fix for it in #38, but it should be improved to be a bit smarter I think.

**Comment by TradeFace** [comment]
All done and released under v1.1.7. Good work guys! :100: :1st_place_medal:
